code/modules/mob/living/living.dm

```diff
@@ -426,6 +426,40 @@
 
 		set_pull_offsets(M, state)
 
+/// Adds pass flags and signal to smash through windows as you fly through the air
+/mob/living/proc/start_window_flight(duration = 1.5 SECONDS)
+	RegisterSignal(src, COMSIG_MOVABLE_MOVED, PROC_REF(flying_window_smash))
+	passwindow_on(src, TRAIT_MOVE_FLYING)
+	addtimer(CALLBACK(src, PROC_REF(end_window_flight)), duration)
+
+/// Removes pass flags and signal to smash through windows as you fly through the air
+/mob/living/proc/end_window_flight()
+	UnregisterSignal(src, COMSIG_MOVABLE_MOVED)
+	passwindow_off(src, TRAIT_MOVE_FLYING)
+
+/// Fly through window panes, smashing them on the way
+/mob/living/proc/flying_window_smash(atom/movable/mover, atom/oldloc, direction)
+	SIGNAL_HANDLER
+	var/mob/living/flying_mob = mover
+	var/turf/target_turf = get_turf(flying_mob)
+	for(var/obj/structure/tram/tram_wall in target_turf)
+		tram_wall.deconstruct(disassembled = FALSE)
+		flying_mob.balloon_alert_to_viewers("smashed through!")
+		flying_mob.apply_damage(damage = rand(5, 15), damagetype = BRUTE, wound_bonus = 15, bare_wound_bonus = 25, sharpness = SHARP_EDGED, attack_direction = get_dir(tram_wall, oldloc))
+		new /obj/effect/decal/cleanable/glass(get_step(flying_mob, flying_mob.dir))
+
+	for(var/obj/structure/window/window in target_turf)
+		window.deconstruct(disassembled = FALSE)
+		flying_mob.balloon_alert_to_viewers("smashed through!")
+		flying_mob.apply_damage(damage = rand(5, 15), damagetype = BRUTE, wound_bonus = 15, bare_wound_bonus = 25, sharpness = SHARP_EDGED, attack_direction = get_dir(window, oldloc))
+		new /obj/effect/decal/cleanable/glass(get_step(flying_mob, flying_mob.dir))
+
+	for(var/obj/machinery/door/window/windoor in target_turf)
+		windoor.deconstruct(disassembled = FALSE)
+		flying_mob.balloon_alert_to_viewers("smashed through!")
+		flying_mob.apply_damage(damage = rand(5, 15), damagetype = BRUTE, wound_bonus = 15, bare_wound_bonus = 25, sharpness = SHARP_EDGED, attack_direction = get_dir(windoor, oldloc))
+		new /obj/effect/decal/cleanable/glass(get_step(flying_mob, flying_mob.dir))
```

Comment:
This can certainly be accomplished without copy pasting the same block of code thrice

Reply:
I'll have a look

Reply:
Yeah, it doesn't need windoor anyways, can be a proc on structure.